### PR TITLE
Workflow for building & testing git-annex on Windows

### LINKED ISSUE #37: Build (test) git-annex on Windows in github actions
https://git-annex.branchable.com/install/Windows/ provides some instructions on HOWTO.  If they aren't sufficient - we will dig deeper, e.g. ask @joeyh and may be Yuri who established Jenkins setup in qemu on one of the servers in the dungeons, but it is a black box to me.

Note:
- ATM windows builds come without libmagic/file support.  So enabling that one could be postponed until basic build setup is functioning. Some pointers etc from my attempts could be found at [git-annex website comment](https://git-annex.branchable.com/todo/provide_windows_build_with_MagicMime/#comment-36e31d0ae3d55820482ce7e35badd70c)

probably should be a separate workflow to start with, and not sure if it would be feasible to meld it into the one we have for linux + OSX now.

### PR DESCRIPTION
**Pull request by jwodder**
Closes #37.

### CODE REVIEW

**Comment by jwodder** [comment]
@yarikoptic Status: The build is failing when trying to link git-annex.exe with the error message "realgcc.exe: error: CreateProcess: No such file or directory".  Research indicates that this is caused by too-long file paths and can be solved by placing the repository at a shorter path, yet it continues to happen even when cloning the git-annex repository into C:\Temp\git-annex.

Compiling on the Windows 7 VM you sent me works fine; it's only on GitHub Actions that I'm seeing this happen.

**Comment by yarikoptic** [comment]
ok, thanks.  I will look/ask around.  Meanwhile, while in "crack this Windows nut" mode -- could you please try building (e.g. locally) [libmagic](https://git-annex.branchable.com/todo/provide_windows_build_with_MagicMime/#comment-36e31d0ae3d55820482ce7e35badd70c) with haskell bindings and then building annex against it?

**Comment by yarikoptic** [comment]
@jwodder - pardon a basic question:
- how come `stack` comes into that environment, or it comes with it? -- I do not see any installation for stack, just its invocation as it was always there (it is not somehow coming with scoop, NSIS - please extend comments what are those for for ignorant ones like me)

Re failure:
- can we have `stack` (installed? moved?) under even shorter `C:\sr` (now as suggested e.g. in https://github.com/haskell/haskell-ide-engine/issues/1778 which suggests also some changes to `%STACK_ROOT%\config.yaml` (I see no STACK_ROOT tuneups etc).
- ghc + github related issue I found  is https://github.com/actions/virtual-environments/issues/1052 which has a snippet to double-check if long paths are enabled in the particular environment we are using here.
  - From there I reached the build recipe for haskell itself: https://github.com/haskell/haskell-language-server/blob/master/.github/workflows/build.yml#L20  -- so may be some hints there (shorter executable names etc) would be of relevance?

**Comment by jwodder** [comment]
@yarikoptic 
* Stack comes pre-installed in the Windows environments on GitHub Actions, as listed [here](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md).  I don't think it can be easily moved.
* The git-annex build scripts need [NSIS](https://en.wikipedia.org/wiki/Nullsoft_Scriptable_Install_System) in order to create a Windows installer.  My original research pointed to using scoop (some sort of Windows package manager) as the best way to install NSIS in a GitHub Actions environment, but I now see in the link above that NSIS comes pre-installed in those environments.

**Comment by yarikoptic** [comment]
ran into https://github.com/actions/setup-haskell  -- have you considered that one -  may be it "incorporates all great best practices" for building on Windows envs on github workflows?

**Comment by yarikoptic** [comment]
>  I don't think it can be easily moved.

but another instance could be installed I guess (unless may be that setup-haskell just does it all "nicely"?)

**Comment by yarikoptic** [comment]
woohooo -- it is green for building and artifact is there! awesome. So it is failing due to tests -- most likely due to abbreviated long path, and/or too long of a path (less likely).  If there is a way to set TMPDIR to point somewhere like `c:\temp` - it might provide remedy

```
  Remote Tests
    remote type directory
  Detected a filesystem without fifo support.
  Disabling ssh connection caching.
  Detected a crippled filesystem.
  Entering an adjusted branch where files are unlocked as this filesystem does not support locked files.
Switched to branch 'adjusted/master(unlocked)'
      init:                                               FAIL
        Exception: MoveFileEx "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\ran9C18" Just ".git\\annex\\objects\\627\\ccf\\SHA256E-s1048576--c29db6e3e6a9911d559753b424694c883d4b65e5e1f1eee74ed58bc496907460.this-is-a-test-key\\SHA256E-s1048576--c29db6e3e6a9911d559753b424694c883d4b65e5e1f1eee74ed58bc496907460.this-is-a-test-key": does not exist (The system cannot find the path specified.)
      unavailable remote
        removeKey:                                        
Error: The operation was canceled.
```

**Comment by yarikoptic** [comment]
might be of interest to @joeyh since likely it is also windows-specific "shortnened" path specific gotcha yet to be address.

**Comment by yarikoptic** [comment]
Oh hoh, seems like we are just step away from the success - it just came find .mgc magic database file

**Comment by jwodder** [comment]
@yarikoptic I don't know why it can't.  The output from the "DEBUG" step shows that `magic.mgc` is located in `/usr/bin`, the same directory as `libmagic-1.dll`, and [file's code](https://github.com/file/file/blob/3810fb8881131bae5b4c4c2acb0b5464b35ff1ea/src/magic.c#L222) seems to indicate that it looks next to the dll (among other locations) for `magic.mgc`.

**Comment by yarikoptic** [comment]
IIRC last time we succeeded to figure out where it is looking by running command via strace in cygwin terminal within VM.  Could you try downloading/installing produced installer, and running in cygwin via strace?

**Comment by jwodder** [comment]
@yarikoptic `/c/Program\ Files/Git/usr/bin/git-annex version` in both git-bash and Cygwin (with `/cygdrive` prepended) on my VM immediately returns 127 without printing anything, and prepending `strace` gives an error.

**Comment by yarikoptic** [comment]
ehm, while trying I am getting just a pop up gui box with `Procedure entry point could not be located in the dynamic link library msvcrt.dll` error (like [here](https://answers.microsoft.com/en-us/windows/forum/windows_7-performance/procedure-entry-point-could-not-be-located-in-the/66982e1f-c7ae-44d3-84a3-fd07578801bb)).  First I thought it is because I have 32bit Git and we have 64 bit build of annex here.  So I uninstalled Git and git-annex completely, installed fresh Git and then git-annex, but error persists... I wonder why me and why you don't see a similar error in your case on local VM... if you saw it too - then may be it is because we use "windows latest" here while that VM is Windows 7.  but since not -- need to figure out...

**Comment by yarikoptic** [comment]
I am not yet speaking the "lingua" so cannot tell exactly the source of the problem, but I suspect newer windows or mingw used on build host... I will push a commit to try building on `windows-2016` instead of `windows-latest` and see where it would get us (if anywhere)

**Comment by yarikoptic** [comment]
the most relevant discussion seems to be here: https://github.com/simonmichael/hledger/issues/1039

**Comment by yarikoptic** [comment]
so [here](https://github.com/simonmichael/hledger/issues/1039#issuecomment-500299293) : I built commit 9b3fc06 on Win7 using a 32-bit Haskell Stack and nightly-2019-06-10 as the resolver in stack.yaml. hledger-web runs without problems there.

git-annex's stack.yaml has `resolver: lts-16.16`, which seems to be recent enough if I take it corresponds to the ones from https://www.stackage.org/ and which had
```
LTS 16.19 for ghc-8.8.4, published 3 days ago
LTS 16.11 for ghc-8.8.3, published a month ago
```
so I guess 16.16 is somewhere in between, and git-annex switched to it just on Oct 2, so it cannot be it... may be you understand better the situation there @jwodder and @joeyh ?

**Comment by yarikoptic** [comment]
some notes/thoughts out loud:

blindly copying msvcrt.dll from the build machine didn't help: although it does provide that dreaded `_create_locale`, even if I move original msvcrt.dll to e.g. `c:\tmp`, windows seems to manage to find it instead of the one I placed, and does not pick up from current directory -- some kind of "magic" to me. And under cygwin `ldd` lists 
```
DataLad@datalads-win7 /cygdrive/c/Program Files/Git/usr/bin
$ ldd /cygdrive/c/Windows/System32/msvcrt.dll
        ntdll.dll => /cygdrive/c/Windows/SYSTEM32/ntdll.dll (0x77490000)
        kernel32.dll => /cygdrive/c/Windows/system32/kernel32.dll (0x77270000)
        KERNELBASE.dll => /cygdrive/c/Windows/system32/KERNELBASE.dll (0x7fefd0c0000)
        ??? => ??? (0x7fef1580000)
```
so may be that newer one needs one more dll (not sure how to figure that out without having some newer windows available).

All roads for such an error and haskell seems to lead eventually to https://github.com/kazu-yamamoto/unix-time/issues/57 which is still open (since if I got it right no solution which fits all was found), I wonder if that is the ultimate issue - but I do not see mentioning of any unix-time in the [log](https://pipelines.actions.githubusercontent.com/2UPlDxaVvvbkeFX4btxWorCjpJvj40zvWY5ogH2yZibhOMcU7O/_apis/pipelines/1/runs/1852/signedlogcontent/3?urlExpires=2020-10-22T01%3A09%3A19.4315726Z&urlSigningMethod=HMACV1&urlSignature=n54RplCRUQQx49e5zxaIAWudbyMhAwhCBLUy%2BHfImI0%3D)...

also how does `actions/setup-haskell@v1.1` which installs ghc latest (8.10.1) "works" with stack spec for `resolver: lts-16.16` which should be ghc 8.8.3 (both are later than 8.6.5 which was [told](https://github.com/simonmichael/hledger/issues/1039#issuecomment-501812690) to avoid need for patched unix-time, so likely unix-time has nothing to do with this here), which is installed later -- but which one is actually used? well, stack says "This will not interfere with any system-level installation", so hopefully that is the case but does it matter and why we have two of them?

so it seems all boiling down to "stop using no longer supported windows 7".  I will look into installing a freshier version tomorrow and give a shot there

**Comment by yarikoptic** [comment]
TL;DR:  https://docs.microsoft.com/en-us/sysinternals/downloads/procmon is our killer tool for such debugging

No total victory yet (midnight, need to have some lunch and get to bed), but progress.  I have instantiated devel VM of windows 10 from https://developer.microsoft.com/en-us/windows/downloads/virtual-machines/ (also added to https://github.com/datalad-tester/dev-containers/tree/master/windows but datalad would still first download that monster .zip and then extract).  Installed git (64 bit) and then git-annex from win2016 build -- it runs and I reproduced your finding of not finding magic db:
```
C:\tmp\test>git annex add  lua5.1.dll                                                                                                                                                             git-annex: bad annex.largefiles configuration in gitattributes: Parse failure: unable to load magic database; "mimeencoding" cannot be used 
```
then used procmon filtering on magic and got

```
8:39:18.3421697 PM	git-annex.exe	5224	CreateFile	C:\Users\User\AppData\Local\.magic\magic.mgc	PATH NOT FOUND	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
8:39:18.3425349 PM	git-annex.exe	5224	CreateFile	C:\Users\User\Local Settings\Application Data\.magic\magic.mgc	REPARSE	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: <unknown>
8:39:18.3426378 PM	git-annex.exe	5224	CreateFile	C:\Users\User\AppData\Local\Application Data\.magic\magic.mgc	REPARSE	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: <unknown>
8:39:18.3427355 PM	git-annex.exe	5224	CreateFile	C:\Users\User\AppData\Local\.magic\magic.mgc	PATH NOT FOUND	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
8:39:18.3429318 PM	git-annex.exe	5224	CreateFile	C:\Program Files\Common Files\.magic\magic.mgc	PATH NOT FOUND	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
8:39:18.3434135 PM	git-annex.exe	5224	CreateFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a, OpenResult: Opened
8:39:18.3435718 PM	git-annex.exe	5224	QueryBasicInformationFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	CreationTime: 10/21/2020 8:12:27 PM, LastAccessTime: 10/21/2020 8:34:50 PM, LastWriteTime: 10/19/2020 8:43:22 PM, ChangeTime: 10/21/2020 8:12:27 PM, FileAttributes: A
8:39:18.3436008 PM	git-annex.exe	5224	CloseFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	
8:39:18.3438123 PM	git-annex.exe	5224	CreateFile	C:\Program Files\Git\usr\share\misc\magic.mime.mgc	NAME NOT FOUND	Desired Access: Read Attributes, Disposition: Open, Options: Open Reparse Point, Attributes: n/a, ShareMode: Read, Write, Delete, AllocationSize: n/a
8:39:18.3438980 PM	git-annex.exe	5224	CreateFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	Desired Access: Generic Read, Disposition: Open, Options: Synchronous IO Non-Alert, Non-Directory File, Attributes: N, ShareMode: Read, Write, AllocationSize: n/a, OpenResult: Opened
8:39:18.3439731 PM	git-annex.exe	5224	QueryInformationVolume	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	VolumeCreationTime: 9/14/2020 11:00:02 AM, VolumeSerialNumber: AEDE-FB18, SupportsObjects: True, VolumeLabel: 
8:39:18.3439863 PM	git-annex.exe	5224	QueryAllInformationFile	C:\Program Files\Git\usr\share\misc\magic.mgc	BUFFER OVERFLOW	CreationTime: 10/21/2020 8:12:27 PM, LastAccessTime: 10/21/2020 8:34:50 PM, LastWriteTime: 10/19/2020 8:43:22 PM, ChangeTime: 10/21/2020 8:12:27 PM, FileAttributes: A, AllocationSize: 6,656,000, EndOfFile: 6,652,192, NumberOfLinks: 1, DeletePending: False, Directory: False, IndexNumber: 0x10000000463a0, EaSize: 0, Access: Generic Read, Position: 0, Mode: Synchronous IO Non-Alert, AlignmentRequirement: Word
8:39:18.3477802 PM	git-annex.exe	5224	ReadFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	Offset: 0, Length: 6,652,192, Priority: Normal
8:39:18.3546370 PM	git-annex.exe	5224	CloseFile	C:\Program Files\Git\usr\share\misc\magic.mgc	SUCCESS	
```

where it shows that it does find magic.mgc in C:\Program Files\Git\usr\share\misc\magic.mgc	 but then there are some other failures and BUFFER OVERFLOW... :-/  Then reran while filtering for process `git-annex` and got the following dump worth looking/grepping through I guess - full dump of events which came from git-annex process: http://www.onerussian.com/tmp/procmon-git-annex.txt

**Comment by yarikoptic** [comment]
I knew that I must not give up now.  I **won**.  Summary: I stared at above lines for a bit and realized that it didn't find that magic.mgc -- it created it instead under `C:\Program Files\Git\usr\share\misc\magic.mgc` !!!   So I went and copied original one you added to .exe distribution from bin into that share/misc -- and IT WORKED!!!
```
C:\tmp\test>git annex add  lua5.1.dll                                                                                                                                                             add lua5.1.dll                                                                                                                                                                                    ok                                                                                                                                                                                                (recording state in git...)    

C:\tmp\test>git annex add test2                                                                                                                                                                   add test2 (non-large file; adding content to git repository) ok                                                                                                                                   (recording state in git...)                                                                                                                                                                                                    
```
for the test though we need to call `git annex sync` so it creates a commit in `master` since on this crippled file system it operates in adjusted branch.

**Comment by jwodder** [comment]
@yarikoptic The PR is finally ready to merge.

**Comment by yarikoptic** [comment]
in my windows 10 devel VM, uninstalled git-annex, installed from recent artifact here (note: needed to login to see it), installed, added binary and text -- all worked out nicely!  Great!

**Comment by joeyh** [comment]
This is some impressive work.

Are there any plans to drop a build artifact out of this, that could be distributed to users?

**Comment by joeyh** [comment]
The problem with cloning should be fixed now.

**Comment by jwodder** [comment]
@joeyh Unfortunately, cloning is still broken on Windows, this time because some of the filenames are too long.  ([See comment on git-annex issue](https://git-annex.branchable.com/bugs/Error_cloning_repository_on_Windows/#comment-516a4d3ed683a4f43b8c240aac60b89f).)
